sphinxext/code_ref.py

```diff
@@ -0,0 +1,332 @@
+"""Sphinx extension providing formatted code blocks, referencing some function."""
+
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, Literal, get_args
+
+from docutils import nodes
+from docutils.parsers.rst import directives
+from sphinx.util.docutils import SphinxDirective
+from sphinx.util.parsing import nested_parse_to_nodes
+
+from altair.vegalite.v5.schema._typing import VegaThemes
+from tools.codemod import embed_extract_func_def, extract_func_def
+
+if TYPE_CHECKING:
+    import sys
+    from typing import (
+        Any,
+        Callable,
+        ClassVar,
+        Iterable,
+        Iterator,
+        Mapping,
+        Sequence,
+        TypeVar,
+        Union,
+    )
+
+    from docutils.parsers.rst.states import RSTState, RSTStateMachine
+    from docutils.statemachine import StringList
+    from sphinx.application import Sphinx
+
+    if sys.version_info >= (3, 12):
+        from typing import TypeAliasType
+    else:
+        from typing_extensions import TypeAliasType
+    if sys.version_info >= (3, 10):
+        from typing import TypeAlias
+    else:
+        from typing_extensions import TypeAlias
+
+    T = TypeVar("T")
+    OneOrIter = TypeAliasType("OneOrIter", Union[T, Iterable[T]], type_params=(T,))
+
+_OutputShort: TypeAlias = Literal["code", "plot"]
+_OutputLong: TypeAlias = Literal["code-block", "altair-plot"]
+_OUTPUT_REMAP: Mapping[_OutputShort, _OutputLong] = {
+    "code": "code-block",
+    "plot": "altair-plot",
+}
+_Option: TypeAlias = Literal["output", "fold", "summary"]
+
+_PYSCRIPT_URL_FMT = "https://pyscript.net/releases/{0}/core.js"
+_PYSCRIPT_VERSION = "2024.10.1"
+_PYSCRIPT_URL = _PYSCRIPT_URL_FMT.format(_PYSCRIPT_VERSION)
+
+
+def validate_output(output: Any) -> _OutputLong:
+    output = output.strip().lower()
+    if output not in {"plot", "code"}:
+        msg = f":output: option must be one of {get_args(_OutputShort)!r}"
+        raise TypeError(msg)
+    else:
+        short: _OutputShort = output
+        return _OUTPUT_REMAP[short]
+
+
+def validate_packages(packages: Any) -> str:
+    if packages is None:
+        return '["altair"]'
```

Comment:
## Note
I tested replacing with [.../altair-5.4.1-py3-none-any.whl](https://files.pythonhosted.org/packages/9b/52/4a86a4fa1cc2aae79137cc9510b7080c3e5aede2310d14fae5486feec7f7/altair-5.4.1-py3-none-any.whl)

Had some intermittent errors, but when it worked it was a **lot** faster since `pandas`, `numpy` are so heavy

![image](https://github.com/user-attachments/assets/61177b61-71c3-4b88-b4b7-9f7cbf1e19d9)

As @mattijn [noted](https://github.com/vega/altair/pull/3630#issuecomment-2408718433), if we could get `pyodide` to update to `altair>=5.4.0` this would speed things up and be easier to maintain

Reply:
You knows how that works @joelostblom? Ref https://github.com/pyodide/pyodide/pull/4580.

Reply:
I found this yesterday while looking for how it works
- https://github.com/pyodide/pyodide/issues/4918

Reply:
Need to mark as resolved to merge, but it would probably be good to track this somewhere